Add S3 has-policy-permission filter with support for wildcards in bucket policies

blt04 opened the pull request:
This adds a new `has-policy-permission` S3 filter.  The filter looks for specific S3 permissions (actions and principals) in a S3 bucket policy.  It works even when bucket policies have wildcards in actions.

For example, it can be hard to determine if a bucket policy grants write access with existing filters (such as `has-statement`) because actions can have wildcards.  For example, any of the following actions can grant write access to a bucket:

* `s3:*`
* `s3:Put*`
* `s3:PutObject`

By using the `has-policy-permission` filter, users can specify the permission(s) they are looking for and this filter will match if the bucket policy grants that permission.

Example filter:

```
policies:
  - name: s3-public-writable-policy
    resource: s3
    filters:
    - type: has-policy-permission
      actions:
        - s3:PutObject
        - s3:PutObjectAcl
        - s3:PutBucketPolicy
      principals:
        - AWS:*
```

The filter matches if any specified s3 actions are allowed by the bucket policy _and_ any of the specified principals are allowed by the policy.

kapilt commented:
thanks for the pr, definitely want to add a unit test for this one.

kapilt commented on c7n/resources/s3.py:
```diff
@@ -783,6 +785,165 @@ def process_bucket(self, b):
+    Special considerations:
+
+    * Principals should be specified as "prefix:value" where prefix is one of AWS,
+      Federated, Service, etc.
```
nice docs, i'd probably say reference principal in iam documentation via url here.

kapilt commented on c7n/resources/s3.py:
```diff
@@ -783,6 +785,165 @@ def process_bucket(self, b):
+
+        return None
+
+    def __get_action_matches(self, statement, required_actions):
```
one underscore is fine to denote private, class mangling from double __ is typically gratuitous and just as easy to bypass.

blt04 commented:
Thanks for the review.  I addressed some of the comments in the latest commit.

I don't quite understand the `cross-account` filter comparison.  The main strength of the filter I'd like to add, `has-policy-permission`, is that it makes it easy to scan for permissions even if the bucket policy has actions with wildcards.  I found this was hard to do with existing filters.

I think of this as analogous to how the security groups filters have special processing for ports.  A security group might whitelist ports 20-25, but specifying `Ports: [22]` in a c7n policy will still match.  Similarly, with this S3 filter, a bucket policy might allow the `s3:Put*` wildcard action, but specifying `s3:PutObject` in a c7n policy will still match.

Thoughts?

kapilt commented:
the `cross-account` filter comparison, is because it also is checking bucket policies, and supports action wild cards of the same form here, so i'm trying to understand the use case here to see if it would be better to support directly there if its not already supported.

ie. this policy should also match on buckets w policies granting access to principals outside of the account even if the policy is using wild card syntax for the statement's action.

```
policies:
   - name: find-s3-buckets-with-perm-wildcards
      resource: s3
      filters:
        - type: cross-account
           actions: 
             - s3:PutObject
             - s3:PutObjectAcl
             - s3:PutBucketPolicy
```

the main benefit for cross-account is that its a common filter across all resources with embedded resources policies.

the primary question is if the use case difference for this pr is about allowing intra account principals to be specified, if thats the case it might be better to extend cross-account to allow for the same. Or if there's a bucket policy that you think it should be matching and its not then and it needs a bug fix/additional test case.

kapilt commented:
ugh.. sorry did approval on wrong pr :-(

kapilt commented:
ping, any comments on comparison to `cross-account`?

kapilt commented:
@blt04 ping ^?

blt04 commented:
Sorry for the delay.  We can close this.  I did a quick comparison and `cross-account` works fine.  There are some differences that cause me to get different results but they are mainly around conditions.

I didn't realize cross-account existed when I wrote this.
